### DISCUSSION TITLE
yarn berry not being used if you follow the documentation

### DISCUSSION THREAD

**Discussion by darkbasic** [mikro-orm/mikro-orm#4143] [General]
https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md#submitting-a-pull-request

According to your documentation `corepack enable` followed by `yarn` should be enough to get started.
Unfortunately that way you end up using the deprecated yarn v1.
To use yarn v3 you will have to run `yarn set version berry` first:
```
➤ YN0000: Retrieving https://repo.yarnpkg.com/3.4.1/packages/yarnpkg-cli/bin/yarn.js
➤ YN0000: Saving the new release in .yarn/releases/yarn-3.4.1.cjs
➤ YN0000: Done in 0s 810ms
```
which also sets the yarnPath:
```
diff --git a/.yarnrc.yml b/.yarnrc.yml
index 3186f3f0..fc8ec93a 100644
--- a/.yarnrc.yml
+++ b/.yarnrc.yml
@@ -1 +1,3 @@
 nodeLinker: node-modules
+
+yarnPath: .yarn/releases/yarn-3.4.1.cjs
```

Also you gitignored `.yarn/plugins` and `.yarn/releases` which contain the Yarn releases used in the current repository (as defined by yarn set version): https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored

**Comment by B4nan** [comment]
You might need to do `corepack prepare yarn@stable --activate`, that's all. Will add that to the guide.

> Also you gitignored .yarn/plugins and .yarn/releases which contain the Yarn releases used in the current repository (as defined by yarn set version): [yarnpkg.com/getting-started/qa#which-files-should-be-gitignored](https://yarnpkg.com/getting-started/qa#which-files-should-be-gitignored)

I did, on purpose, as I don't want to commit yarn binaries to the repository. Using corepack is alternative to that, there is no point in using corepack if you commit them.

From [yarn docs](https://yarnpkg.com/getting-started/install):

> The preferred way to manage Yarn is through [Corepack](https://nodejs.org/dist/latest/docs/api/corepack.html), a new binary shipped with all Node.js releases starting from 16.10. It acts as an intermediary between you and Yarn, and lets you use different package manager versions across multiple projects **without having to check-in the Yarn binary anymore**.

**Reply by darkbasic** [reply]
> You might need to do corepack prepare yarn@stable --activate, that's all. Will add that to the guide.

Uhm... I've already tried it (saw it in the CI) but it didn't work.

**Reply by darkbasic** [reply]
```
niko@talos2 ~/devel/mikro-orm $ node --version
v18.15.0
niko@talos2 ~/devel/mikro-orm $ npm -g list --depth=0
/home/niko/.nvm/versions/node/v18.15.0/lib
├── corepack@0.15.3
└── npm@9.5.0
```

**Reply by darkbasic** [reply]
```
niko@talos2 ~/devel/mikro-orm $ corepack prepare yarn@stable --activate
Preparing yarn@stable for immediate activation...
niko@talos2 ~/devel/mikro-orm $ yarn --version
1.22.19
```

Outside of the mikro-orm directory:
```
niko@talos2 ~ $ yarn --version
3.4.1
```

**Reply by B4nan** [reply]
Maybe you have the `yarn` alias in PATH that points to yarn v1 or something like that? It does work fine for me, it works in the CI, and we adopted this recently in the company I work with, also works for my colleagues there. 

FYI for me it also automatically switches versions based on the package.json's `packageManager`. I even get a 404 when I try to define nonexisting yarn version in there and use yarn afterwards.

<img width="1243" alt="image" src="https://user-images.githubusercontent.com/615580/225726592-94b04bb9-6ad9-4af5-8e27-172cbcd47c5a.png">

**Reply by B4nan** [reply]
> Outside of the mikro-orm directory:

That's really weird.

**Reply by darkbasic** [reply]
Yes I have multiple paths pointing to different versions of yarn (system wide for Gentoo ebuilds consumption and the nvm one), but it's weird that outside of the mikro-orm repo it always end up using the nvm version while somehow inside the project directory it skips to v1.
I have no idea what is going on and why but it's clearly specific to my configuration and should be easy enough to workaround it, I don't think there is any need to update the documentation (unless `corepack prepare yarn@stable --activate` is really needed).

**Reply by B4nan** [reply]
> niko@talos2 ~/devel/mikro-orm $ npm -g list --depth=0

Hmm, maybe its because you have corepack installed via NPM? It comes bundled with node 18, so maybe try removing that and run `corepack enable` again?

**Reply by darkbasic** [reply]
I didn't install it, it ends up there when I install node.js via nvm. nvm used to install it even before node 18, so it's probably still installing it and ends up messing somehow. No idea though.

**Reply by darkbasic** [reply]
```
niko@talos2 ~/devel/mikro-orm $ nvm --version
0.39.1
```

It's not too old but maybe it's worth updating nvm as well at this point.

**Reply by B4nan** [reply]
Maybe switch to `fnm` right ahead, it's much faster (no lag on opening new terminal).

https://github.com/Schniz/fnm

**Reply by darkbasic** [reply]
Good idea, I was too lazy to check out if it builds on ppc64le but it should probably be fine and I've had enough headaches with nvm anyway.

**Reply by darkbasic** [reply]
Turned out fnm does indeed **NOT** work on ppc64le, but I've managed to find an [IBM branch of ring](https://github.com/briansmith/ring/pull/1057#issuecomment-1296454035) which compiles fine, at least if you [keep around your .git directory](https://github.com/gentoo/guru/pull/90). fnm is still somehow broken because it thinks to be on amd64 and it [downloads x64 binaries](https://github.com/Schniz/fnm/issues/942) but the workaround is easy enough at this point.
The good news is that now it works like it should:

```
niko@talos2 ~ $ whereis yarn
yarn: /usr/bin/yarn

niko@talos2 ~ $ corepack enable

niko@talos2 ~ $ whereis yarn
yarn: /usr/bin/yarn /home/niko/.local/share/fnm/node-versions/v18.15.0/installation/bin/yarn

niko@talos2 ~ $ yarn --version
3.4.1
```

Unfortunately once you enable yarn 3 via `corepack enable` it forces yarn berry onto every yarn legacy project. Do you know how to avoid such behaviour?

**Reply by B4nan** [reply]
I believe you can use `corepack prepare yarn@1 --activate` outside of a project directory.

<img width="349" alt="image" src="https://user-images.githubusercontent.com/615580/226648059-c96476f7-1abb-4148-89b2-499d239db0f9.png">

**Reply by darkbasic** [reply]
That did the trick, but now the following vscode task fails whenever I work on the mikro-orm code base:
```
Jest (mikro-orm)

[info] auto config:

found:
  rootPath: /home/niko/devel/mikro-orm
  jestCommandLine: yarn test


error @mikro-orm/root@5.0.0: The engine "yarn" is incompatible with this module. Expected version ">=3.2.0". Got "1.22.19"
error Commands cannot run with an incompatible environment.


[error] failed to retrieve test file list. TestExplorer might show incomplete test items
```
